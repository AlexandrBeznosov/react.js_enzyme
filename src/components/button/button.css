.button-style {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    width: 200px;
    height: 50px;
    background-color: orange;
    color: white;
    border-radius: 10px;
    margin: 10px;
    cursor: pointer;
}