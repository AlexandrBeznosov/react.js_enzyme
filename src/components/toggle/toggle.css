.toggle-style {
    margin: 10px;
    cursor: pointer;
}