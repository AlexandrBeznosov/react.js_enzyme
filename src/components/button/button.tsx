import { useCallback } from "react";

import './button.css';

export interface ButtonProps {
    title: string;
    isDisabled: boolean;
    onClick: (message: string) => void;
}

export const Button: React.FC<ButtonProps> = props => {
    const { onClick } = props;

    const handleClick = useCallback(() => {
        if (!props.isDisabled) {
            onClick('Hello!');
        }
    }, [onClick, props.isDisabled]);

    return <div className="button-style" data-test-target="button-element" onClick={handleClick}>{props.title}</div>
}