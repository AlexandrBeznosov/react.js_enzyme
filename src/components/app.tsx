import React from 'react';

import { Button } from './button/button';
import { Toggle } from './toggle/toggle';

const App: React.FC = () => {
    return (<div className="App">
        <Button title='Click me!' onClick={message => window.alert(message)} isDisabled={false} />
        <Toggle title='Toggle me!' onChange={state => window.console.log(String(state))} initialState={true} />
    </div>);
}

export { App };
