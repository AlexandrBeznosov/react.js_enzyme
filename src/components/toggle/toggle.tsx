import { useCallback, useState } from "react";

import './toggle.css';

export interface ToggleProps {
    title: string;
    initialState: boolean;
    onChange: (state: boolean) => void;
}

export const Toggle: React.FC<ToggleProps> = props => {
    const { onChange } = props;

    const [state, setState] = useState(props.initialState);

    const handleChange = useCallback(() => {
        setState(!state);
        onChange(!state);

    }, [onChange, state]);

    return (
        <label className="toggle-style" htmlFor="toggle-id">
            <input type="checkbox" checked={state} id="toggle-id" data-test-target="toggle-element" onChange={handleChange} />
            {props.title}
        </label>
    );
}