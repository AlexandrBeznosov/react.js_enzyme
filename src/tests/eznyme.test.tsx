import { mount, shallow } from 'enzyme';
import React from 'react';
import { App } from '../components/app';
import { Button, ButtonProps } from '../components/button/button';
import { Toggle, ToggleProps } from '../components/toggle/toggle';

describe('Enzyme do:', () => {
    test('create snapshot of mounted component', () => {
        const instance = mount(<App />);

        expect(instance).toMatchSnapshot();
    });

    test('find child by component', () => {
        const instance = shallow(<App />);
        const button = instance.find(Button);

        expect(button).toHaveLength(1);
    });

    test('find child by attribute', () => {
        const app = mount(<App />);
        const button = app.find('[data-test-target="button-element"]');

        expect(button).toHaveLength(1);
    });

    test('find child by class name', () => {
        const props: ToggleProps = {
            initialState: true,
            title: 'Dummy toggle title',
            onChange: () => null,
        }
        const toggle = mount(<Toggle {...props} />);
        const input = toggle.find('.toggle-style');

        expect(input).toHaveLength(1);
    });

    test('get child attribute', () => {
        const app = mount(<App />);
        const input = app.find(Toggle).find('input');
        input.simulate('change');
        const isChecked = input.prop('checked');

        expect(isChecked).toBeTruthy();
    });

    test('set component property', () => {
        const props: ButtonProps = {
            title: 'Dummy button title',
            isDisabled: false,
            onClick: jest.fn(),
        }
        const button = shallow(<Button {...props} />);
        button.setProps({ ...props, isDisabled: true });
        button.simulate('click');

        expect(props.onClick).toBeCalledTimes(0);
    });

    test('callback test', () => {
        const props: ButtonProps = {
            title: 'Dummy button title',
            isDisabled: false,
            onClick: jest.fn(),
        }
        const button = shallow(<Button {...props} />);
        button.simulate('click')

        expect(props.onClick).toBeCalledTimes(1);
    });
});

