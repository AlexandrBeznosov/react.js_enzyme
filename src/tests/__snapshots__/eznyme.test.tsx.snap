// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enzyme do: create snapshot of mounted component 1`] = `
<App>
  <div
    className="App"
  >
    <Button
      isDisabled={false}
      onClick={[Function]}
      title="Click me!"
    >
      <div
        className="button-style"
        data-test-target="button-element"
        onClick={[Function]}
      >
        Click me!
      </div>
    </Button>
    <Toggle
      initialState={true}
      onChange={[Function]}
      title="Toggle me!"
    >
      <label
        className="toggle-style"
        htmlFor="toggle-id"
      >
        <input
          checked={true}
          data-test-target="toggle-element"
          id="toggle-id"
          onChange={[Function]}
          type="checkbox"
        />
        Toggle me!
      </label>
    </Toggle>
  </div>
</App>
`;
